drivers/adodb-ldap.inc.php

```diff
@@ -32,54 +32,95 @@
 }
 
 class ADODB_ldap extends ADOConnection {
-	var $databaseType = 'ldap';
-	var $dataProvider = 'ldap';
-
-	# Connection information
-	var $username = false;
-	var $password = false;
-
-	# Used during searches
-	var $filter;
-	var $dn;
-	var $version;
-	var $port = 389;
+	public $databaseType = 'ldap';
+	public $dataProvider = 'ldap';
+		
+	/*
+	* Caches the version info
+	*/
+	protected $version;
+	
 
-	# Options configuration information
-	var $LDAP_CONNECT_OPTIONS;
+	/*
+	* Options configuration information
+	* @deprecated 5.23
+	*/
+	public $LDAP_CONNECT_OPTIONS;
 
 	# error on binding, eg. "Binding: invalid credentials"
-	var $_bind_errmsg = "Binding: %s";
+	protected $_bind_errmsg = "Binding: %s";
 
-	// returns true or false
+	protected $connectionParameters = array(
+		LDAP_OPT_PROTOCOL_VERSION=>3,
+		LDAP_OPT_REFERRALS=>0
+	);
 
-	function _connect( $host, $username, $password, $ldapbase)
+	
+	/**
+	 * Connect to a AD Server
+	 *
+	 * @param string|null $ldapServer The AD Server to connect to.
+	 * @param string|null $username The username to connect as.
+	 * @param string|null $password The password to connect with.
+	 * @param string|null $ldapbase The Base DN
+	 *
+	 * @return bool|null True if connected successfully, false if connection failed, or null if the mysqli extension
+	 * isn't currently loaded.
+	 */
+	public function _connect( $ldapServer, $username, $password, $ldapbase)
 	{
 		global $LDAP_CONNECT_OPTIONS;
 
-		if ( !function_exists( 'ldap_connect' ) ) return null;
-
-		if (strpos($host,'ldap://') === 0 || strpos($host,'ldaps://') === 0) {
-			$this->_connectionID = @ldap_connect($host);
-		} else {
-			$conn_info = array( $host,$this->port);
-
-			if ( strstr( $host, ':' ) ) {
-				$conn_info = explode( ':', $host );
-			}
+		if ( !function_exists( 'ldap_connect' ) ) 
+			return null;
 
-			$this->_connectionID = @ldap_connect( $conn_info[0], $conn_info[1] );
+		
+		if (strpos($ldapServer,'ldap://') !== 0 && strpos($ldapServer,'ldaps://') !== 0) 
+		{
+			/*
+			* If ldap SSL not explicitly specified, fall back to insecure
+			*/ 
+			$ldapServer = sprintf('ldap://%s',$ldapServer);
 		}
-		if (!$this->_connectionID) {
-			$e = 'Could not connect to ' . $conn_info[0];
+
+		
+		$this->_connectionID = @ldap_connect($ldapServer);
+		
+		if (!$this->_connectionID) 
+		{
+			$e = 'Could not connect to ' . $ldapServer;
 			$this->_errorMsg = $e;
-			if ($this->debug) ADOConnection::outp($e);
+			if ($this->debug) 
+				ADOConnection::outp($e);
+			
 			return false;
 		}
-		if( count( $LDAP_CONNECT_OPTIONS ) > 0 ) {
+
+		if( is_array($LDAP_CONNECT_OPTIONS) && count( $LDAP_CONNECT_OPTIONS ) > 0 ) 
+		{
+			/*
```

Comment:
Should we output a deprecation warning here ? e.g. 
```php
trigger_error('$LDAP_CONNECT_OPTIONS is deprecated, use setConnectionParameter() instead', E_USER_DEPRECATED);
```

Reply:
changed, per suggestion